Create New Game mode - CHOOSE

jriyyya opened the pull request:
For this game mode the player has to choose from the cats displayed on the screen which has a odd count.

edit : Added another type in this game mode, which now ask the player to choose the cat on the screen which has an even count.

@sourabhaa @chimosky Please Review

jriyyya committed:
Add new gametype CHOOSE in declartion of  GameType

jriyyya committed:
Initialize properties of gameArea class

- choose_option_cats is the array of options
  available to be selected in the CHOOSE
  gamemode

- chosen_cat is the cat which is selected by the
  player to be correct

- odd_cat_id is the id of the cat which is correct
  answer for CHOOSE gameMode and has a odd count

jriyyya committed:
Add method __define_choose_option_cats ro GameArea

- this function is used to define the array we initialised

- It calculates the coordinates for the options dynamically

- It appends the 4 options to the choose_option_cats array

jriyyya committed:
Modify function __draw_selected_option

- An if-else condition is added according to the gameType currently
being played, as now this function is used in gameType CHOOSE as-well

- For the gameType CHOOSE, when player hovers on any cats available as
an option to select from, an hover indicator is shown around that cat

jriyyya committed:
made a function __draw_choose_options

- This function calls __define_choose_option_cats() which defines cats

- we then call draw method of each of these cats

jriyyya committed:
Detect motion on the cats placed for Choosing

- chosen_cat is assigned the value of the hover cat

jriyyya committed:
Assign selected_option as chosen_cat on press

jriyyya committed:
Extend _release_cb functionality

- Now it support both GameType.ROWS
  and GameType.CHOOSE

jriyyya committed:
Add Game over & start message for gameType CHOOSE

jriyyya committed:
Made a function divide_into_even_odd

- This function takes in a number, it then makes an array of 4
number, whose sum is that number, such that 1 of the number in the array is
odd and rest all are even

- it then flattens the above array, so that it is converted into an
array of cat ids. the count of each cat id is the same as
the entry at that position in the
divided_nums array

jriyyya committed:
Appending cats in array cat

- The functionality of gameType Choose is added in the function add_cat

- the individual cat count is taken from the array which is returned by the
function divide_into_even_odd

jriyyya committed:
Added levels of new GameType in Json

chimosky commented on gamearea.py:
You could merge this with the commit where `__draw_choose_options` is created.

chimosky commented on gamearea.py:
```diff
@@ -268,7 +268,7 @@ def __release_cb(self, widget, event):
         if "type" not in self.level_data.keys():
             return 0
-        if self.level_data["type"] == GameType.ROWS:
+        if self.level_data["type"] == GameType.ROWS or self.level_data["type"] == GameType.CHOOSE:
```
The commit message says " Add functionlity of _release_cb in gameType CHOOSE" but the change in the commit doesn't reflect that.

jriyyya replied:
Yes, I realised it was quite confusing.

I have changed the commit message [7f629b2](https://github.com/sugarlabs/classify-cats/pull/9/commits/7f629b24a274f77578c2e64c7cef418a725eb857)

chimosky commented on gamearea.py:
```diff
@@ -441,8 +442,21 @@ def __draw_end_message(self, context):
+        elif self.level_data["type"] == GameType.CHOOSE:
+            if self.selected_option is not None:
+                if self.selected_option.cat_id == self.odd_cat_id:
+                    message = _("You Choose correctly!")
+                    self.win = True
+                else:
+                    message = _("You Choose wrong")
+                    self.win = False
+            else:
+                message = _("You should have chosen an option")
```
choose should be small letters and it should be in past tense as the time this message shows, you're expected to have made your choices.

chimosky replied:
This still uses "Choose" and not "chose".

chimosky replied:
Seen your change.

chimosky commented on gamearea.py:
```diff
@@ -511,6 +511,25 @@ def show_message(self, context, message, font_size, y=0):
+        divided_nums.append(remaining_sum)
+        divided_nums[self.odd_cat_id - 1] += 1
+
+        for i in range(0, 4):
+            for j in range(0, divided_nums[i]):
```
You can omit the 0's in this loop.

chimosky replied:
Any reason why you still have the range starting from 0?

chimosky commented:
Reviewed 5a7dc44, not tested.

jriyyya committed:
made choose to past tense in messages

jriyyya committed:
Avoid using 0s in range starting index

jriyyya commented:
@chimosky I have made the suggested changes

chimosky commented on gamearea.py:
```diff
@@ -237,6 +237,15 @@ def __motion_cb(self, widget, event):
+        elif self.level_data["type"] == GameType.CHOOSE:
+            flag_hover = False
+            for cat in self.choose_option_cats:
+                if event.x > cat.x and event.x < cat.x + cat.width and event.y > cat.y and event.y < cat.y + cat.height:
```
This could be made more readable as it looks confusing.

jriyyya replied:
Done

chimosky commented on gamearea.py:
```diff
@@ -546,18 +546,24 @@ def add_cats(self):
+            cat_width = 60
+            cat_height = 60
 
+        for x in range(0, cats):
```
Another range starting with 0.

jriyyya committed:
fix instance of using 0 in range starting index

jriyyya committed:
refactor collision detection for better readabilty

chimosky commented on gamearea.py:
```diff
@@ -240,7 +240,7 @@ def __motion_cb(self, widget, event):
             flag_hover = False
             for cat in self.choose_option_cats:
-                if event.x > cat.x and event.x < cat.x + cat.width and event.y > cat.y and event.y < cat.y + cat.height:
+                if (cat.x < event.x < cat.x + cat.width) and (cat.y < event.y < cat.y + cat.height):
```
This isn't that readable, a better check would be;
```python3
if event.x > cat.x and event.x < (cat.x + cat.width) and event.y > cat.y and event.y < (cat.y + cat.height):
```

jriyyya committed:
refactor collision detection again for better readabilty

chimosky commented:
Tested 863f284, works as expected although the time it takes to start the choose game mode is quite short considering the text that's displayed on the screen and sometimes it makes it a bit difficult to read the instruction for the game mode.

jriyyya committed:
Made changes so that instruction screen is more readable
- increased the time given to read the instruction
  from 3 seconds to 5 seconds

- increased the font size of the instruction so
  that it gets the player's attention first

jriyyya commented:
<img width="640" alt="image" src="https://github.com/sugarlabs/classify-cats/assets/96080203/9c558857-bc8f-43ac-82df-590b8c785689">


@chimosky Is it more readable now?

chimosky committed:
Review - change instruction

Signed-off-by: Chihurumnaya Ibiam <ibiam@sugarlabs.org>

chimosky commented:
> @chimosky Is it more readable now?

Yes it is.

I've made a small change, it'll be great to add an event count mode too, not just odd.

jriyyya commented:
So I should continue this PR, or can I create a new PR for this?

jriyyya commented:
So should I continue this PR, or can I create a new PR for this?

chimosky commented:
> So should I continue this PR, or can I create a new PR for this?

This is for the new game mode so it makes sense to add a part of the said game mode here.

jriyyya commented:
> This is for the new game mode so it makes sense to add a part of the said game mode here.

Okay, I will do so

jriyyya committed:
Add even type for gamemode CHOOSE
- Before the game was to choose the cat
  which has an odd count

- Now the game also supports the type in
  which the even count has to be chosen

- choose_type in class initialization of gamearea
  has a random value of 0 or 1, where 0 is for
  even game type and 1 is for odd game type

- divide_into_even_odd function is made such that
  it first divide the number of cats into 4 cats
  type, all of even count, Then in gametype even
  3 cats type are made odd, by increasing their
  count by 1 and in odd gametype, one cat type
  count is increased by 1

- in gametype even, 0 cats of any type is considered
  even, so while playing that game, if 0 cat of any
  type is present, that cat has to be chosen from
  options to win that round

jriyyya committed:
removed print statements

chimosky commented on gamearea.py:
```diff
@@ -522,7 +531,16 @@ def divide_into_even_odd(self, number):
             remaining_sum -= even_number
         divided_nums.append(remaining_sum)
-        divided_nums[self.odd_cat_id - 1] += 1
+        print(divided_nums)
```
I'm guessing you forgot to remove this after debugging.

jriyyya replied:
Yes, Thank you.

Also I was trying out something before and forgot to remove some parts of it and committed it by mistake, removed them as well.

chimosky commented on gamearea.py:
```diff
@@ -522,7 +531,16 @@ def divide_into_even_odd(self, number):
+                if (i + 1) != self.choose_cat_id:
+                    divided_nums[i] += 1
+
+        print(divided_nums, self.choose_type)
```
This too.

jriyyya committed:
removed buggy code

chimosky commented:
I tried testing and I never encountered the odd game mode but I did notice something odd;

[Screencast from 2023-09-01 18-05-41.webm](https://github.com/sugarlabs/classify-cats/assets/16203834/820e1d12-b768-4771-91c7-1c01bd4538d3)

As you'll notice from the screencast, I fail the question after choosing correctly, I chose blue which was even and failed.
Noticing this for the first time hence bringing it up and it seems it happens when they're two cats that fall into the category of the question - not shown in the video above - .

jriyyya commented:
The game gets over after some specific level count, and at the end of the screen the final score is shown, I think it added up the score to the final score, so you didn't fail the final round.

Also I was using random to choose between 0 and 1 for even and odd game mode, very strange that you weren't able to encounter the odd game mode 😅

chimosky commented:
> The game gets over after some specific level count, and at the end of the screen the final score is shown, I think it added up the score to the final score, so you didn't fail the final round.

We do need a help button to explain that as it's not obvious, my experience is about the game play and not the end. 

> Also I was using random to choose between 0 and 1 for even and odd game mode, very strange that you weren't able to encounter the odd game mode 😅

How was 4d6b81c buggy?

jriyyya commented:
> We do need a help button to explain that as it's not obvious, my experience is about the game play and not the end. 

I will look into it.

> How was 4d6b81c buggy?

I was thinking about merging the win check condition for game type row and choose, and later realised it is kind of not a good idea, and I forgot to remove some of it's code

jriyyya commented:
Hi @chimosky, Fixed it.

jriyyya committed:
Bug fix in how cats were divided in even and odd
 - Before the total number of cats to be placed
   on screen was greater than the number of cats
   defined for that round in levels.json because
   of which few cats were not getting rendered

jriyyya commented:
@chimosky Can you please merge this, We can then make a release for this.

chimosky commented:
Tested 47c8b9a, works as expected. 

We can look at making the game not end where it does, it's quite short.

Thanks for working on this!

jriyyya commented:
@chimosky How about we make the max puzzle count as 10 so that the game doesn't end soon, currently it's 5
